File: ServeurCompteDepot/services/TransfertService.cs
using ServeurCompteDepot.Models;
using ServeurCompteDepot.models;
using Microsoft.EntityFrameworkCore;

namespace ServeurCompteDepot.Services
{
    public interface ITransfertService
    {
        Task<IEnumerable<Transfert>> GetAllTransfertsAsync();
        Task<Transfert?> GetTransfertByIdAsync(int id);
        Task<IEnumerable<Transfert>> GetTransfertsByCompteAsync(string idCompte);
        Task<IEnumerable<Transfert>> GetTransfertsByCompteEnvoyeurAsync(string idCompteEnvoyeur);
        Task<IEnumerable<Transfert>> GetTransfertsByCompteReceveurAsync(string idCompteReceveur);
        Task<Transfert> CreateTransfertAsync(string compteEnvoyeur, string compteReceveur, decimal montant, DateTime dateTransfert);
        Task<decimal> GetTotalTransfertsSortantsAsync(string compteId);
        Task<decimal> GetTotalTransfertsEntrantsAsync(string compteId);
        Task<int> GetNombreTransfertsSortantsAsync(string compteId);
        Task<int> GetNombreTransfertsEntrantsAsync(string compteId);
        // Nouvelles méthodes avec frais
        Task<IEnumerable<TransfertAvecFrais>> GetAllTransfertsAvecFraisAsync();
        Task<IEnumerable<TransfertAvecFrais>> GetTransfertsByCompteAvecFraisAsync(string idCompte);
    }

    public class TransfertService : ITransfertService
    {
        private readonly CompteDepotContext _context;
        private readonly ITransactionService _transactionService;
        private readonly IHistoriqueSoldeService _historiqueSoldeService;
        private readonly IFraisService _fraisService;

        public TransfertService(CompteDepotContext context, ITransactionService transactionService, IHistoriqueSoldeService historiqueSoldeService, IFraisService fraisService)
        {
            _context = context;
            _transactionService = transactionService;
            _historiqueSoldeService = historiqueSoldeService;
            _fraisService = fraisService;
        }

        public async Task<IEnumerable<Transfert>> GetAllTransfertsAsync()
        {
            return await _context.Transferts
                .Include(t => t.CompteEnvoyeur)
                .Include(t => t.CompteReceveur)
                .OrderByDescending(t => t.DateTransfert)
                .ToListAsync();
        }

        public async Task<Transfert?> GetTransfertByIdAsync(int id)
        {
            return await _context.Transferts
                .Include(t => t.CompteEnvoyeur)
                .Include(t => t.CompteReceveur)
                .FirstOrDefaultAsync(t => t.IdTransfert == id);
        }

        public async Task<IEnumerable<Transfert>> GetTransfertsByCompteAsync(string idCompte)
        {
            return await _context.Transferts
                .Include(t => t.CompteEnvoyeur)
                .Include(t => t.CompteReceveur)
                .Where(t => t.Envoyer == idCompte || t.Receveur == idCompte)
                .OrderByDescending(t => t.DateTransfert)
                .ToListAsync();
        }

        public async Task<IEnumerable<Transfert>> GetTransfertsByCompteEnvoyeurAsync(string idCompteEnvoyeur)
        {
            return await _context.Transferts
                .Include(t => t.CompteReceveur)
                .Where(t => t.Envoyer == idCompteEnvoyeur)
                .OrderByDescending(t => t.DateTransfert)
                .ToListAsync();
        }

        public async Task<IEnumerable<Transfert>> GetTransfertsByCompteReceveurAsync(string idCompteReceveur)
        {
            return await _context.Transferts
                .Include(t => t.CompteEnvoyeur)
                .Where(t => t.Receveur == idCompteReceveur)
                .OrderByDescending(t => t.DateTransfert)
                .ToListAsync();
        }

        public async Task<Transfert> CreateTransfertAsync(string compteEnvoyeur, string compteReceveur, decimal montant, DateTime dateTransfert)
        {
            using var transaction = await _context.Database.BeginTransactionAsync();
            
            try
            {
                // Vérifier que les comptes existent
                var compteEnv = await _context.Comptes.FirstOrDefaultAsync(c => c.IdCompte == compteEnvoyeur);
                var compteRec = await _context.Comptes.FirstOrDefaultAsync(c => c.IdCompte == compteReceveur);
                
                if (compteEnv == null)
                    throw new InvalidOperationException($"Compte envoyeur {compteEnvoyeur} introuvable");
                
                if (compteRec == null)
                    throw new InvalidOperationException($"Compte receveur {compteReceveur} introuvable");
                
                // Vérifier que le solde est suffisant (pas de découvert autorisé pour les comptes dépôt)
                if (compteEnv.Solde < montant)
                    throw new InvalidOperationException($"Solde insuffisant. Solde actuel: {compteEnv.Solde}, Montant demandé: {montant}");
                
                // Créer la transaction sortante (débit)
                var transactionSortante = new Transaction
                {
                    IdCompte = compteEnvoyeur,
                    IdTypeTransaction = 4, // Virement sortant
                    Montant = montant,
                    DateTransaction = DateTime.SpecifyKind(dateTransfert, DateTimeKind.Utc)
                };
                
                _context.Transactions.Add(transactionSortante);
                await _context.SaveChangesAsync();
                
                // Créer la transaction entrante (crédit)
                var transactionEntrante = new Transaction
                {
                    IdCompte = compteReceveur,
                    IdTypeTransaction = 3, // Virement entrant
                    Montant = montant,
                    DateTransaction = DateTime.SpecifyKind(dateTransfert, DateTimeKind.Utc)
                };
                
                _context.Transactions.Add(transactionEntrante);
                await _context.SaveChangesAsync();
                
                // Créer le transfert
                var transfert = new Transfert
                {
                    DateTransfert = dateTransfert.Date, // Utiliser seulement la date
                    IdTransactionEnvoyeur = transactionSortante.IdTransaction.ToString(),
                    IdTransactionReceveur = transactionEntrante.IdTransaction.ToString(),
                    Montant = montant,
                    Envoyer = compteEnvoyeur,
                    Receveur = compteReceveur
                };
                
                _context.Transferts.Add(transfert);
                await _context.SaveChangesAsync();
                
                // Mettre à jour les soldes
                compteEnv.Solde -= montant;
                compteRec.Solde += montant;
                
                // Créer les historiques de solde
                var histoSoldeEnvoyeur = new HistoriqueSolde
                {
                    IdCompte = compteEnvoyeur,
                    IdTransaction = transactionSortante.IdTransaction,
                    Montant = compteEnv.Solde,
                    DateChangement = DateTime.SpecifyKind(dateTransfert, DateTimeKind.Utc)
                };
                
                var histoSoldeReceveur = new HistoriqueSolde
                {
                    IdCompte = compteReceveur,
                    IdTransaction = transactionEntrante.IdTransaction,
                    Montant = compteRec.Solde,
                    DateChangement = DateTime.SpecifyKind(dateTransfert, DateTimeKind.Utc)
                };
                
                _context.HistoriquesSolde.Add(histoSoldeEnvoyeur);
                _context.HistoriquesSolde.Add(histoSoldeReceveur);
                
                await _context.SaveChangesAsync();
                await transaction.CommitAsync();
                
                return transfert;
            }
            catch
            {
                await transaction.RollbackAsync();
                throw;
            }
        }

        public async Task<decimal> GetTotalTransfertsSortantsAsync(string compteId)
        {
            return await _context.Transferts
                .Where(t => t.Envoyer == compteId)
                .SumAsync(t => t.Montant);
        }

        public async Task<decimal> GetTotalTransfertsEntrantsAsync(string compteId)
        {
            return await _context.Transferts
                .Where(t => t.Receveur == compteId)
                .SumAsync(t => t.Montant);
        }

        public async Task<int> GetNombreTransfertsSortantsAsync(string compteId)
        {
            return await _context.Transferts
                .Where(t => t.Envoyer == compteId)
                .CountAsync();
        }

        public async Task<int> GetNombreTransfertsEntrantsAsync(string compteId)
        {
            return await _context.Transferts
                .Where(t => t.Receveur == compteId)
                .CountAsync();
        }

        /// <summary>
        /// Récupère tous les transferts avec les frais associés
        /// </summary>
        public async Task<IEnumerable<TransfertAvecFrais>> GetAllTransfertsAvecFraisAsync()
        {
            var transferts = await GetAllTransfertsAsync();
            return await EnrichirTransfertsAvecFraisAsync(transferts);
        }

        /// <summary>
        /// Récupère les transferts d'un compte avec les frais associés
        /// </summary>
        public async Task<IEnumerable<TransfertAvecFrais>> GetTransfertsByCompteAvecFraisAsync(string idCompte)
        {
            var transferts = await GetTransfertsByCompteAsync(idCompte);
            return await EnrichirTransfertsAvecFraisAsync(transferts);
        }

        /// <summary>
        /// Enrichit une liste de transferts avec les informations sur les frais
        /// </summary>
        private async Task<IEnumerable<TransfertAvecFrais>> EnrichirTransfertsAvecFraisAsync(IEnumerable<Transfert> transferts)
        {
            var transfertsAvecFrais = new List<TransfertAvecFrais>();

            foreach (var transfert in transferts)
            {
                var transfertAvecFrais = new TransfertAvecFrais(transfert);

                try
                {
                    // Récupérer la transaction de sortie (virement sortant)
                    if (!string.IsNullOrEmpty(transfert.IdTransactionEnvoyeur) && 
                        int.TryParse(transfert.IdTransactionEnvoyeur, out int idTransactionEnvoyeur))
                    {
                        var transactionEnvoyeur = await _context.Transactions
                            .FirstOrDefaultAsync(t => t.IdTransaction == idTransactionEnvoyeur);

                        if (transactionEnvoyeur != null)
                        {
                            // Récupérer les frais applicables pour cette transaction de type "Virement sortant"
                            var frais = await _fraisService.FindCurrentFraisAsync(
                                "Virement sortant",
                                transfert.Montant,
                                transactionEnvoyeur.DateTransaction
                            );

                            if (frais != null)
                            {
                                // Le montant des frais est directement la valeur du frais
                                var montantFrais = frais.Valeur;

                                transfertAvecFrais.FraisEnvoyeur = montantFrais;
                                transfertAvecFrais.LibelleFraisEnvoyeur = frais.Nom;
                                transfertAvecFrais.MontantTotalEnvoyeur = transfert.Montant + montantFrais;
                            }
                            else
                            {
                                // Pas de frais applicables
                                transfertAvecFrais.FraisEnvoyeur = 0;
                                transfertAvecFrais.LibelleFraisEnvoyeur = "Aucun frais";
                                transfertAvecFrais.MontantTotalEnvoyeur = transfert.Montant;
                            }
                        }
                    }

                    // Si pas de transaction trouvée, initialiser avec des valeurs par défaut
                    if (string.IsNullOrEmpty(transfertAvecFrais.LibelleFraisEnvoyeur))
                    {
                        transfertAvecFrais.FraisEnvoyeur = 0;
                        transfertAvecFrais.LibelleFraisEnvoyeur = "Non calculé";
                        transfertAvecFrais.MontantTotalEnvoyeur = transfert.Montant;
                    }
                }
                catch (Exception ex)
                {
                    // En cas d'erreur, initialiser avec des valeurs par défaut
                    Console.WriteLine($"Erreur lors du calcul des frais pour le transfert {transfert.IdTransfert}: {ex.Message}");
                    transfertAvecFrais.FraisEnvoyeur = 0;
                    transfertAvecFrais.LibelleFraisEnvoyeur = "Erreur calcul";
                    transfertAvecFrais.MontantTotalEnvoyeur = transfert.Montant;
                }

                transfertsAvecFrais.Add(transfertAvecFrais);
            }

            return transfertsAvecFrais;
        }
    }
}

File: ServeurCompteCourant/src/main/java/com/example/serveurcomptecourant/services/TransfertService.java
package com.example.serveurcomptecourant.services;

import java.math.BigDecimal;
import java.util.List;
import java.util.logging.Level;
import java.util.logging.Logger;

import com.example.serveurcomptecourant.exceptions.CompteCourantBusinessException;
import com.example.serveurcomptecourant.exceptions.CompteCourantException;
import com.example.serveurcomptecourant.models.CompteCourant;
import com.example.serveurcomptecourant.models.Transfert;
import com.example.serveurcomptecourant.models.TransfertAvecFrais;
import com.example.serveurcomptecourant.models.Transaction;
import com.example.serveurcomptecourant.models.Frais;
import com.example.serveurcomptecourant.repository.TransfertRepository;
import com.example.serveurcomptecourant.repository.CompteCourantRepository;
import com.example.serveurcomptecourant.repository.TransactionRepository;

import jakarta.ejb.EJB;
import jakarta.ejb.Stateless;
import jakarta.ejb.TransactionAttribute;
import jakarta.ejb.TransactionAttributeType;

@Stateless
public class TransfertService {
    private static final Logger LOGGER = Logger.getLogger(TransfertService.class.getName());
    
    @EJB
    private TransfertRepository transfertRepository;
    
    @EJB
    private CompteCourantRepository compteRepository;
    
    @EJB
    private TransactionRepository transactionRepository;
    
    @EJB
    private FraisService fraisService;

    /**
     * Récupère tous les transferts
     */
    public List<Transfert> getAllTransferts() throws CompteCourantException {
        try {
            return transfertRepository.findAll();
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Erreur lors de la récupération des transferts", e);
            throw new CompteCourantException("Erreur lors de la récupération des transferts", e);
        }
    }

    /**
     * Récupère les transferts d'un compte avec les frais associés
     */
    public List<TransfertAvecFrais> getTransfertsByCompteAvecFrais(String compteId) throws CompteCourantException {
        try {
            List<Transfert> transferts = getTransfertsByCompte(compteId);
            return enrichirTransfertsAvecFrais(transferts);
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Erreur lors de la récupération des transferts avec frais du compte " + compteId, e);
            throw new CompteCourantException("Erreur lors de la récupération des transferts avec frais", e);
        }
    }
    
    /**
     * Récupère tous les transferts avec les frais associés
     */
    public List<TransfertAvecFrais> getAllTransfertsAvecFrais() throws CompteCourantException {
        try {
            List<Transfert> transferts = getAllTransferts();
            return enrichirTransfertsAvecFrais(transferts);
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Erreur lors de la récupération de tous les transferts avec frais", e);
            throw new CompteCourantException("Erreur lors de la récupération des transferts avec frais", e);
        }
    }
    
    /**
     * Enrichit une liste de transferts avec les informations sur les frais
     */
    private List<TransfertAvecFrais> enrichirTransfertsAvecFrais(List<Transfert> transferts) throws CompteCourantException {
        List<TransfertAvecFrais> transfertsAvecFrais = new java.util.ArrayList<>();
        
        for (Transfert transfert : transferts) {
            TransfertAvecFrais transfertAvecFrais = new TransfertAvecFrais(transfert);
            
            try {
                // Récupérer la transaction de sortie (virement sortant)
                if (transfert.getIdTransactionEnvoyeur() != null) {
                    Long idTransactionEnvoyeur = Long.parseLong(transfert.getIdTransactionEnvoyeur());
                    Transaction transactionEnvoyeur = transactionRepository.findById(idTransactionEnvoyeur.intValue());
                    
                    if (transactionEnvoyeur != null) {
                        // Récupérer les frais applicables pour cette transaction de type "Virement sortant"
                        Frais frais = fraisService.findCurrentFrais(
                            "Virement sortant",
                            transfert.getMontant(),
                            transactionEnvoyeur.getDateTransaction()
                        );
                        
                        if (frais != null) {
                            // Calculer le montant des frais
                            BigDecimal montantFrais = BigDecimal.valueOf(frais.getValeur());
                            
                            transfertAvecFrais.setFraisEnvoyeur(montantFrais);
                            transfertAvecFrais.setLibelleFraisEnvoyeur(frais.getNom());
                            transfertAvecFrais.setMontantTotalEnvoyeur(transfert.getMontant().add(montantFrais));
                        } else {
                            // Pas de frais applicables
                            transfertAvecFrais.setFraisEnvoyeur(BigDecimal.ZERO);
                            transfertAvecFrais.setLibelleFraisEnvoyeur("Aucun frais");
                            transfertAvecFrais.setMontantTotalEnvoyeur(transfert.getMontant());
                        }
                    }
                }
                
                // Si pas de transaction trouvée, initialiser avec des valeurs par défaut
                if (transfertAvecFrais.getFraisEnvoyeur() == null) {
                    transfertAvecFrais.setFraisEnvoyeur(BigDecimal.ZERO);
                    transfertAvecFrais.setLibelleFraisEnvoyeur("Non calculé");
                    transfertAvecFrais.setMontantTotalEnvoyeur(transfert.getMontant());
                }
                
            } catch (Exception e) {
                LOGGER.log(Level.WARNING, "Erreur lors du calcul des frais pour le transfert " + transfert.getId(), e);
                // En cas d'erreur, initialiser avec des valeurs par défaut
                transfertAvecFrais.setFraisEnvoyeur(BigDecimal.ZERO);
                transfertAvecFrais.setLibelleFraisEnvoyeur("Erreur calcul");
                transfertAvecFrais.setMontantTotalEnvoyeur(transfert.getMontant());
            }
            
            transfertsAvecFrais.add(transfertAvecFrais);
        }
        
        return transfertsAvecFrais;
    }

    /**
     * Récupère les transferts d'un compte (envoyeur ou receveur)
     */
    public List<Transfert> getTransfertsByCompte(String compteId) throws CompteCourantException {
        try {
            if (compteId == null || compteId.trim().isEmpty()) {
                throw new CompteCourantBusinessException("L'ID du compte est obligatoire");
            }
            
            // Vérifier que le compte existe
            CompteCourant compte = compteRepository.find(compteId);
            if (compte == null) {
                throw new CompteCourantBusinessException.CompteNotFoundException(0L);
            }
            
            return transfertRepository.findByCompte(compteId);
        } catch (CompteCourantBusinessException e) {
            throw e;
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Erreur lors de la récupération des transferts du compte " + compteId, e);
            throw new CompteCourantException("Erreur lors de la récupération des transferts", e);
        }
    }
    /**
     * Crée un transfert (appelé par TransactionService)
     */
    @TransactionAttribute(TransactionAttributeType.REQUIRED)
    public Transfert createTransfert(String compteEnvoyeur, String compteReceveur, BigDecimal montant, 
                                   Transaction transactionSortante, Transaction transactionEntrante) throws CompteCourantException {
        try {
            // Validation des paramètres
            validateTransfertData(compteEnvoyeur, compteReceveur, montant, transactionSortante, transactionEntrante);
            
            // Créer le transfert
            Transfert transfert = new Transfert();
            transfert.setMontant(montant);
            transfert.setEnvoyer(compteEnvoyeur);
            transfert.setReceveur(compteReceveur);
            transfert.setIdTransactionEnvoyeur(transactionSortante.getId().toString());
            transfert.setIdTransactionReceveur(transactionEntrante.getId().toString());
            transfert.setDateTransfert(transactionSortante.getDateTransaction().toLocalDate());
            
            return transfertRepository.save(transfert);
            
        } catch (CompteCourantBusinessException e) {
            throw e;
        } catch (Exception e) {
            LOGGER.log(Level.SEVERE, "Erreur lors de la création du transfert", e);
            throw new CompteCourantException("Erreur lors de la création du transfert", e);
        }
    }
    /**
     * Validation des données de transfert
     */
    private void validateTransfertData(String compteEnvoyeur, String compteReceveur, BigDecimal montant,
                                     Transaction transactionSortante, Transaction transactionEntrante) throws CompteCourantBusinessException {
        if (compteEnvoyeur == null || compteEnvoyeur.trim().isEmpty()) {
            throw new CompteCourantBusinessException("Le compte envoyeur est obligatoire");
        }
        
        if (compteReceveur == null || compteReceveur.trim().isEmpty()) {
            throw new CompteCourantBusinessException("Le compte receveur est obligatoire");
        }
        
        if (montant == null || montant.compareTo(BigDecimal.ZERO) <= 0) {
            throw new CompteCourantBusinessException.MontantInvalideException(
                montant != null ? montant.doubleValue() : 0.0
            );
        }
        
        if (compteEnvoyeur.equals(compteReceveur)) {
            throw new CompteCourantBusinessException("Le compte envoyeur et receveur doivent être différents");
        }
        
        if (transactionSortante == null || transactionEntrante == null) {
            throw new CompteCourantBusinessException("Les transactions associées sont obligatoires");
        }
        
        if (transactionSortante.getId() == null || transactionEntrante.getId() == null) {
            throw new CompteCourantBusinessException("Les transactions doivent être sauvegardées avant de créer le transfert");
        }
    }
}